src/jailer/src/env.rs

```diff
@@ -197,37 +200,53 @@ impl Env {
-        SyscallReturnCode(unsafe { libc::unshare(libc::CLONE_NEWPID) })
-            .into_empty_result()
-            .map_err(Error::UnshareNewPID)?;
-
         // Compute jailer's total CPU time up to the current time.
         self.jailer_cpu_time_us =
             utils::time::get_time_us(utils::time::ClockType::ProcessCpu) - self.start_time_cpu_us;
 
-        // Duplicate the current process. The child process will belong to the previously created
-        // PID namespace.
-        // TODO: replace the `unshare()` + `fork()` combo with `clone()` if we ever need to
-        //  squeeze every bit of start-up latency we can get
-        let pid = unsafe { libc::fork() };
-        match pid {
-            0 => {
-                // Reset process start time.
-                self.start_time_cpu_us = 0;
-
-                Err(Error::Exec(self.exec_command(chroot_exec_file)))
-            }
-            child_pid => {
-                // Save the PID of the process running the exec file provided
-                // inside <chroot_exec_file>.pid file.
-                self.save_exec_file_pid(child_pid, chroot_exec_file)?;
-                unsafe { libc::exit(0) }
-            }
+        // This is a wrapper for our child callback to the clone libc function.
+        // We will use clone's arg parameter as the actual callback and call it from here.
+        extern "C" fn child_fn(arg: *mut libc::c_void) -> libc::c_int {
+            let callback_ref: *mut *mut dyn FnOnce() = arg as *mut *mut dyn std::ops::FnOnce();
+            let callback = unsafe { Box::from_raw(*callback_ref) };
+
+            callback();
+            0
         }
+
+        let child_callback = || {
+            // Reset process start time.
+            self.start_time_cpu_us = 0;
+
+            self.exec_command(chroot_exec_file.clone());
+        };
+
+        let child_callback_box: Box<dyn FnOnce()> = Box::new(child_callback);
+        let child_callback_ref = &mut Box::into_raw(child_callback_box);
+
+        // Mapping the child process it's own stack.
+        let stack: &mut [u8; PROCESS_STACK_SIZE] = &mut [0; PROCESS_STACK_SIZE];
+
+        // The call is safe because we're adding the stack size to itself for using it as stack.
+        let stack_ptr = unsafe { stack.as_mut_ptr().add(stack.len()) };
+
+        // We will use clone in order to spawn a new child process into a new PID namespace.
+        // The current process will not be moved into the newly created namespace, but the child
+        // spawned via clone will assume the role of init(1) in the new namespace.
+        // The call is safe because we're invoking a C library function with valid parameters.
+        let child_pid = unsafe {
+            libc::clone(
+                child_fn,
+                stack_ptr as *mut libc::c_void,
+                libc::SIGCHLD | libc::CLONE_NEWPID,
+                std::mem::transmute(child_callback_ref),
```

Comment:
I think the transmute is ok here, because there is no actual interpretation here, So it is safe. Also, your suggestion will not work and causes segfault.

Reply:
Sorry, I didn't have time to test the suggestions. It causes a segfault, because the cast in `on_clone()` has to be adapted to the new type. What seems to work is the following:

```
        let do_on_clone: &mut dyn FnOnce() = &mut || {
            // Reset process start time.
            self.start_time_cpu_us = 0;

            self.exec_command(chroot_exec_file.clone());
        };

        let do_on_clone_ptr = Box::into_raw(Box::new(do_on_clone));
```

and 

```
        extern "C" fn on_clone(arg: *mut libc::c_void) -> libc::c_int {
            let callback: Box<Box<dyn FnOnce()>> = unsafe { Box::from_raw(arg as _) };

            callback();
            0
        }
```

We need double pointers because trait objects are fat.